### ISSUE TITLE
1/x not the same as inv(x) for matrix in julia 0.2.0

### ISSUE THREAD

**Issue by freefrancisco** [JuliaLang/LinearAlgebra.jl#75]
I was trying to figure out how to invert a matrix in a 0.2.0 installation of julia, and I found out that 1/x gives me a different inverse than inv(x).  
Here is my output

```
julia> x = rand(3, 3)
3x3 Array{Float64,2}:
 0.698453  0.500252  0.822943
 0.483441  0.078515  0.889188
 0.275469  0.341758  0.195685

julia> 1/x
3x3 Array{Float64,2}:
 1.43174   1.99899  1.21515
 2.06851  12.7364   1.12462
 3.63017   2.92605  5.11026

julia> inv(x)
Warning: Possible conflict in library symbol dtrtri_
Warning: Possible conflict in library symbol dgetri_
Warning: Possible conflict in library symbol dgetrf_
3x3 Array{Float64,2}:
  35.4312  -22.5164  -46.69
 -18.4623   11.0545   27.4109
 -17.6333   12.3904   22.9644

julia> versioninfo()
Julia Version 0.2.0
Commit 05c6461 (2013-11-16 23:44 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin12.5.0)
  WORD_SIZE: 64
  BLAS: libgfortblas
  LAPACK: liblapack
  LIBM: libopenlibm

```

**Comment by andreasnoack** [comment]
Because `1` is a number `1/x` gives the reciprocal of each element of the matrix whereas `inv` is the matrix inverse. 

It is better to ask a question like this to the julia-users list.

**Comment by pao** [comment]
[You can find the julia-users list here.](https://groups.google.com/forum/#!forum/julia-users)

**Comment by JeffBezanson** [comment]
I wonder if that behavior is still a good idea. After all `a^-1` is also matrix inverse.

**Comment by jiahao** [comment]
1/A computing the elementwise reciprocal of A is inconsistent with A/B computing B\A. Why not make 1./A compute the elementwise reciprocal instead?

**Comment by freefrancisco** [comment]
Also there is at least one julia tutorial that expects 1/A == A^-1
http://www.statalgo.com/2012/04/15/statistics-with-julia-linear-algebra-with-lapack/

**Comment by jiahao** [comment]
It feels like 1.0 ./A should be the matrix reciprocal, eye()/A should be the matrix inverse and 1.0/A should just be an error.

**Comment by jiahao** [comment]
That tutorial is quite outdated.

**Comment by andreasnoack** [comment]
Agree. I was too fast here. There is also the consequence

```
julia> [1 1]\1
1x2 Array{Float64,2}:
 1.0  1.0

julia> [1 1]\[1]
2-element Array{Float64,1}:
 0.5
 0.5
```

Most linear algebra function accept `Number` interpreted as a `Vector` of length one or a 1x1 `Matrix` so I guess the right thing for `\(Matrix,Number)` is to check if the dimensions are correct.

**Comment by StefanKarpinski** [comment]
If we were to go down the road of making `1/A` compute a matrix inverse, then we'd also want to change the behavior of a lot of other operations. For example, `A + 1` should be equivalent to `A + eye(size(A))` while `A .+ 1` would be how you write the broadcasting sense. I've strongly considered this kind of change before, but it would be pretty disruptive. I think the only path to get there would be to deprecate the non-broadcasting forms for a fairly long time and then reintroduce the new "algebraically correct" behavior.

**Comment by JeffBezanson** [comment]
I would seriously consider doing that. The case of `+` is a bit odd since I don't see that very often, but `/` and `\` are used with matrices routinely and the `1/A` change seems like the right thing.

**Comment by jiahao** [comment]
`A + 1` (or in more conventional notation, `A + λ*I`) is actually a very important use case: `A + λ*I` is used in level shifts for all sorts of matrix operations, particularly ill-conditioned matrix eigenvalue problems; `(A + λI)⁻¹` is a [resolvent kernel](http://en.wikipedia.org/wiki/Resolvent_formalism) for inverting linear operators in the spectral domain; `sqrt(A² + λ²I)⁻¹` and the like show up in [regularization](http://en.wikipedia.org/wiki/Tikhonov_regularization), to name just three use cases. Currently there isn't a more compact way to write  `A + λ*I` other than `A + λ*eye(A)`, and this notation is wasteful in that it initializes a second matrix only to do little other than to scale and add it to something else. So there are nice algebraic reasons why `A + 1` should promote to `A + eye(A)`.

I'm not convinced, however, that this has to be conflated with `1/A` computing the matrix inverse vs. the elementwise reciprocal.

**Comment by jiahao** [comment]
A few thoughts on the specific issue of matrix inverse notation though:
1. The usual admonition is that people should not compute the matrix inverse `A⁻¹` explicitly. There's really no need to make it easier to compute matrix inverses.
2. However, we do use `A/B` as a synonym for `B/A` (or at least, try to.). So if change `1/A` to be the matrix inverse, then `A/1` ought to be consistent with that as well.
3. The real question is whether an expression containing both a matrix and `1` should have the `1` promote to `eye(N)` (the proposed behavior) or `ones(N,N)` with elementwise operations (essentially the current behavior).

**Comment by JeffBezanson** [comment]
I don't think scalar*matrix could do anything but scale the matrix, no matter how we interpret all the above arguments.

**Comment by jiahao** [comment]
Hm, maybe I wasn't being clear. My point is that there are plenty of contexts in which `A + λ` for a matrix `A` and scalar `λ` is a powerful synonym for `A + λ*eye(A)`.

**Comment by jiahao** [comment]
as opposed to the current behavior of `A + λ`, which behaves as a synonym for `A + λ*ones(A)` (and in other such expressions involving matrices and scalars).

**Comment by StefanKarpinski** [comment]
I would be very much in favor of that change. I can't imagine a single mathematician who would not prefer for `A + λ` to be an efficient and expressive way to modify the diagonal. The reason that and the `1/A` thing should be coupled is that they are both about how we treat mathematical operations on a scalar and a matrix – should it broadcast or should it do the algebraically correct thing? What we're doing now is mathematically sketchy but at least it's consistent. If we changed one and not the other then we'd be neither consistent nor correct. I think that changing both is the best thing – then we'll be both consistent and correct.

**Comment by jiahao** [comment]
This proposal, by the way, is very much as its heart about whether a `Matrix` is a linear algebraic object or a two-dimensional container (c.f. JuliaLang/julia#987). The broadcasting behavior is consistent across all n-dimensional arrays, whereas this proposal would special-case behavior for `n=2`.

**Comment by mbauman** [comment]
I like the idea of making scalars behave like they're multiplied by `I` of the "correct size" when interacting with matrices through non-broadcasting operators.  The fact that there's already operators designated to do element-wise broadcasting makes it a bit easier, too.  Addition is well-defended above, and multiplication is free.  Division is tricky, though: what's the correct size for `I` when dividing by non-square matrices?

If `A` is square, then `eye(size(A)...)/A * A = I`, no?  My linear algebra is rusty, but isn't `eye(size(A,2),size(A,1))/A` the choice to make in general?  I think that should behave like `pinv(A)`.

I know I have tons of `A + λ * eye(size(A))` littered through my Matlab code for state estimation and identification.  I think this could be a great change.

**Comment by kmsquire** [comment]
As an aside, the `size(A)` isn't necessary for `ones` or `eye`:

``` julia
julia> ones(A) == ones(size(A))
true

julia> eye(A) == eye(size(A))
ERROR: no method eye((Int64,Int64))

julia> eye(A) == eye(size(A)...)
true
```

**Comment by jiahao** [comment]
@mbauman Tempting as it is to define `1/A` as `pinv(A)` for rectangular matrices, pseudoinverses are left inverses but not right inverses (whereas a true inverse is both a left inverse and a right inverse). So I don't think we can reason consistently about `1/A` for nonsquare matrices in a way that neglects order of operations.

**Comment by jiahao** [comment]
> I know I have tons of A + λ \* eye(size(A)) littered through my Matlab code for state estimation 

You could always define a custom `+(A::Matrix, λ::Number)` method in your code...

**Comment by mbauman** [comment]
Right, of course, since there is no true inverse for rectangular matrices, there is no universal solution here… perhaps the answer is to punt and throw an error for rectangular matrices?  I still like the consistency of (effectively) transforming the scalar `λ` into `λI` across all basic arithmetic operations with matrices.

(Thanks for the tips, Kevin and Jiahao - I'll be happy to ditch those Matlabisms in my Julia code)

**Comment by jiahao** [comment]
@kmsquire thanks, I've edited my comments.

@mbauman we wouldn't have to do anything new at the level of redefining `1/A`, since if it calls `inv(A)`, that would do the dimension check already.

**Comment by carlobaldassi** [comment]
Also, speaking of customization, probably a special matrix type which just keeps a scalar and represents a matrix proportional to the identity of a variable size, together with a constant `I` associated to the value 1, wouldn't be too difficult to write. For example:

```
immutable IdentityMatrix{T<:Number}
    λ::T
end

const I = IdentityMatrix(1)

Base.(:*)(x::Number, M::IdentityMatrix) = IdentityMatrix(x*M.λ)
Base.(:+){T<:Number}(A::Matrix{T}, M::IdentityMatrix) = A + M.λ * eye(A)
Base.(:/){T<:Number}(M::IdentityMatrix, A::Matrix{T}) = M.λ * inv(A)
```

etc. (of course `+` should be written more efficiently).
With these definitions alone one already has:

```
julia> A = rand(2,2)
2x2 Array{Float64,2}:
 0.184807  0.557457
 0.342099  0.889467

julia> A + 2I
2x2 Array{Float64,2}:
 2.18481   0.557457
 0.342099  2.88947 

julia> I/A
2x2 Array{Float64,2}:
 -33.787   21.1754
  12.9948  -7.02  
```

**Comment by StefanKarpinski** [comment]
Yes, that solution had occurred to me as well. It's elegant.

**Comment by jiahao** [comment]
@carlobaldassi the `IdentityMatrix` construct is actually a very strange object, since with a single instantiation `I` you could write things like `(A+I)*(B+I)*(C+I)` regardless of the dimensions of `A`, `B`, `C`, so long as they are such that `A*B*C` is sensible. It's not immediately obvious that it is _wrong_ to define such a thing, but it is certainly not very standard linear algebra to have a single operator defined over multiple vector spaces.

**Comment by StefanKarpinski** [comment]
It is quite common as notation though ;-)

**Comment by jiahao** [comment]
don't be trolling now.

**Comment by jiahao** [comment]
Not to careen too far offpath here, but I've been thinking for awhile that [IterativeSolvers.jl](https://github.com/JuliaLang/IterativeSolvers.jl) really needs a `ShiftedMatrix` type that wraps both the matrix `A` and the shift `λ` and serves as a functional synonym for `A+λI`, so that one need never actually compute `A+λI` explicitly. Instead, one is usually interested in expressions like `(A+λI)\b`, which can be worked out much more efficiently than an explicit construction which may break desirable symmetries or sparsity patterns in `A`.

**Comment by StefanKarpinski** [comment]
Also, it may make sense for `A + I` to be an error if `A` is non-square. In that case all of those `I` are conceptually of the same square shape. Of course, one can concoct other expressions where the `I` are not of the same size.

**Comment by StefanKarpinski** [comment]
The same comment applies to `A + 1` and `1/A` – if these were to do the algebraic thing, it might be best for them to require `A` to be square.

**Comment by jiahao** [comment]
`1/A` should only be defined where `inv(A)` is, i.e. square and nonsingular.

`A+1` can be defined for rectangular `A` trivially by zero-padding the identity matrix with additional columns or rows of zeros as necessary.

**Comment by carlobaldassi** [comment]
It's not written explicitly, but in my sketch above I was actually thinking of only allowing `I` to represent a square matrix (`inv` does assume that of course). To me, the `A + λI` notation seems much more explicit than having a `Number` magically promote to a `Matrix` (BTW I forgot to declare the type as `IdentityMatrix<:AbstractMatrix`); if the latter was the choice though, I'd still support the restriction to square matrices only.

**Comment by stevengj** [comment]
I have to say that I strongly oppose the notation `A + 2` as a synonym for `A + 2*eye(A)`.   For one thing, the only case where this makes any sense at all is for `ndims(A) == 2`, so it would mean that `A + 2` means something completely different depending on the rank of `A`.  ~~For another thing, when you are just manipulating containers of data, adding a scalar to everything in the container is quite a common operation, and this would leave you with no nice, efficient notation for this operation (since `A + 2*ones(A)` is grossly inefficient).~~

For `1/A`, my feelings aren't as strong, but I really don't see a strong motivation to allow this notation for matrix inverses.  My tendency would be to just throw an error in this case, and require the user to either do `1 ./ A` if they want the broadcasting operation or `inv(A)` if they want the inverse.  Trying to guess the user's intention for `1/A` seems like a recipe for bugs.

**Comment by carlobaldassi** [comment]
@stevengj you can always do `A .+ 2`.

**Comment by jiahao** [comment]
Currently, we have that `ones(4,4) + 2 == ones(4,4) .+ 2`. I'm in favor of retaining a set of operations that broadcast scalar updates. However, I do agree that the special-casing of two-dimensional `Array`s is the much more pertinent issue. (I have a comment somewhere above to this effect.)

**Comment by StefanKarpinski** [comment]
We could just make adding a scalar mean to add along the diagonal in N dimensions. I'm just not sure if that's a particularly meaningful operation.

**Comment by stevengj** [comment]
@StefanKarpinski, I don't think that's a commonly useful operation, and it would be an invitation for bugs.   If you go the route of `A + 2 == A + 2eye(A)`, then `+(::Array, ::Number)` should probably not even be defined except for 2d arrays; everything else should be forced to use `.+`.   (If you're going to break a lot of code, you should break it loudly.)   But I have to say I would not be thrilled about this change.

**Comment by jiahao** [comment]
We should think much more carefully about what is meaningful behavior for `N`-dimensional tensors before proceeding with any breaking changes of this magnitude. (c.f. JuliaLang/julia#987, JuliaLang/LinearAlgebra.jl#42; is it sad that I've memorized these issue numbers?)

**Comment by carlobaldassi** [comment]
Small additional comment: if we allowed non-square matrices, we would have things like:

```
julia> a = [0 0 0]
1x3 Array{Int64,2}:
 0  0  0

julia> a + 1
1x3 Array{Int64,2}:
 1  0  0
```

which surely many people would find surprising (I among those).

**Comment by StefanKarpinski** [comment]
Yeah, that's pretty sketchy.

**Comment by jiahao** [comment]
`[0 0 0]` is not a row vector; it is a 1x3 `Matrix`, of rank 1, which mapping of ℝ³ onto ℝ¹, and for which the identity operator (from ℝ³ onto ℝ¹,) has the representation `[1 0 0]`. Not necessarily obvious, but entirely justifiable from an operator viewpoint.

(On a less facetious note, this is why JuliaLang/LinearAlgebra.jl#42 is so thorny.)

**Comment by JeffBezanson** [comment]
Are the changes to `/` and `+` really so tightly coupled? Having `1/A != A^-1` seems like a more serious violation to me.

**Comment by milktrader** [comment]
https://github.com/JuliaStats/TimeSeries.jl/issues/80

I think @jiahao previously mentioned this but _some_ mathematical operators treat for example an `Int` as `int(myvalue(ones(my length)))` but others don't. AFAICT, it applies to `+`, `-`, `*`, `/` but not to `^`

**Comment by JeffBezanson** [comment]
I guess "+, -, *, / are elementwise when one argument is a scalar" is a reasonable and easy-to-remember rule.

**Comment by milktrader** [comment]
Let me elaborate a bit. Given `foo = [1,2]`, it is equivalent to write:

``` julia
julia> foo + 2
2-element Array{Int64,1}:
 3
 4

julia> foo .+ 2
2-element Array{Int64,1}:
 3
 4
```

But not

``` julia
julia> foo .^ 2
2-element Array{Int64,1}:
 1
 4

julia> foo ^ 2
ERROR: no method *(Array{Int64,1}, Array{Int64,1})
 in power_by_squaring at intfuncs.jl:90
 in ^ at intfuncs.jl:113
```

**Comment by milktrader** [comment]
Treating the major four operators (`+`, `-`, `*`, `/`) as special cases where element-wise operations are equivalent to array operations is still is a bit confusing.

``` julia
julia> foo = [1,2]
2-element Array{Int64,1}:
 1
 2

julia> sum(foo, foo)
1-element Array{Int64,1}:
 3

julia> foo + foo
2-element Array{Int64,1}:
 2
 4

julia> foo .+ foo
2-element Array{Int64,1}:
 2
 4
```

**Comment by freefrancisco** [comment]
To me it would make sense that for any mathematical object where a multiplicative identity (left and right) exists that `A op a`  should mean `A op aI` where `A` is the mathematical object in question, `a` is a scalar, and `op` is an operator.  When the identity doesn't exist this operation should be undefined instead of repeating the functionality of the `.op` dotted operators.  

The element-wise operations are well defined already, and the dotted operators are already defined unambiguously to do this.  Why repeat the functionality with the non-dotted operators and waste the opportunity to use them in a different way?  In a way that corresponds to what they actually mean in math notation?  It would also make for clearer code when people want to broadcast to actually use `.op` to show unambiguously that it was their intention.

**Comment by lindahua** [comment]
A + 2 means A + 2 \* ones(size(A)) is a convention that has been well established in technical computing languages and numeric libraries.

Changing this behavior is going to be a huge surprise to everyone, no matter which background he/she comes from. It also probably breaks a huge part of existing codes. 

The existence of .+ is not a strong enough reason to justify such a change. 

I am very strongly against the idea of using A + 2 to express A + 2 I.

**Comment by nalimilan** [comment]
The `A + 2I` really sounds a good idea to me, as it's quite explicit while concise and essentially what you write in math formulas. I don't think any mathematician relies on the convention that `A + 2` means `A + 2I`.

**Comment by freefrancisco** [comment]
@lindahua you make a good argument as to why Matlab can't change what they do, and why they shouldn't.  If the established convention is broken a lot of code will be broken, and a lot of long-time users are going to be displeased.  But if you have a new paradigm that wants to challenge Matlab and be better than what they do, then you have to ask yourself, do I want to do the same things that Matlab does, or do I want to do something different?  

Of course different does not imply better, and sometimes tradition turns out to be the best alternative.  Maybe in this case tradition is good, but maybe in this case tradition is bad.  We don't know, and if we are trying to build a new language that will be better than the traditional languages, then we have to be willing to break with tradition if tradition doesn't make sense.  

Does it make sense in pure mathematics that A+2 means A+2*ones(size(A))?  If it does, then a mathematical language should definitely keep it.  If it doesn't, then a mathematical language should not keep it, regardless of what tradition dictates with other mathematical languages.

**Comment by lindahua** [comment]
Let me elaborate a little bit on my argument.

**First**, it is not just about what MATLAB does, it is a well established semantics adopted by every technical computing language & numerical computing library that I have ever seen. Here is a list:
- `MATLAB`
- `R`
- `Mathematica`
- `NumPy (for Python)`
- `ILNumerics (for C#)` 
- `Eigen (for C++)`
- `Armadillo (for C++)`
-  `NT2 (for C++)` 
- `valarray (C++ standard library)`
  and many more. 

These languages/libraries have different ways to handle arrays (e.g. value semantics, reference semantics, vs. copy-on-write), different ways to differentiate element-wise multiplication and algebraic multiplication. However, they _all_ adopt the same behavior when coming to addition/subtraction between a vector/matrix and a scalar, that is the element-wise behavior.  

One should ask: why this convention is adopted by so many languages & libraries before changing it.

**Second**, in terms of conformance to typical mathematical notations. In every textbook and papers I have read, when people want to express _adding a scalar to the diagonal_, they write `A + λI` instead of `A + λ`. 

From a pure algebraic standpoint, the _Matrix Ring_ is formed of all `n x n` square matrices together with the addition and matrix multiplication defined thereon. Perhaps, I am missing something. I have yet to see an algebraic system (in terms of mathematical concept) that involves addition between matrix and scalar that exhibit the `A + λI` behavior. 

Hence, I haven't seen the justification of this semantics from mathematics.

**Third**, as many others have mentioned, the operation `A + λI` makes sense mostly when `A` is a square matrix. When `A` is a rectangular matrix, a vector, or a cubic array. Such an operation (even you may define it in some sense) is much less useful. However, element-wise operations between such arrays and scalar are still very useful and are actually frequently used. 

Suppose `v` is a vector, then there does exist quite a few papers where the notation `v + λ` is used to express _element-wise_ addition between `v` and `λ`. But I have never seen any body using this to express the weird semantics of adding `λ` to only the first element of `v`.

So if we are trying to enforce this new semantics for matrices, we will be faced with the choice between two very undesirable situation: (1) only enforce this for matrices, and have huge inconsistencies between the behavior of matrices and arrays of other ranks; or (2) enforce the weird & useless behavior to arrays of other ranks (e.g. `v + λ` to mean adding `λ` to only the first element).

**Fourth**, whereas Julia is still a young language. A large amount of codes have been written using Julia, some used in serious projects. Many of these codes relying on the element-wise semantics of `A +  λ` or `v + λ`. The change of this behavior is going to break most of them (sometimes in a very subtle way). Even after we announce this change, most people is going to write codes taking this for granted (as this is so well established as I mentioned above), only to find themselves bited by this unintended change. You may imagine people band their heads against the brick wall when they find that all the subtle bugs are due to `α * v + b` doesn't work as expected and walk away from Julia.

I understand that using `A + λ` to mean `A + λI` may be handy in some situations. But there are much better solution to this (e.g. providing other convenient way to express adding a scalar to diagonal) than changing what `A + λ` means. 

An important principle of interface design is explicit. If you want to express the idea of adding scalars to diagonals, articulate it explicitly (it is reasonable to provide such tools to make it convenient to say that). To change basic semantics that is going to cause huge surprise to everyone and major breakage of existing codes, we need a much stronger reason.

**Comment by goszlanyi** [comment]
+1 for @lindahua's arguments

**Comment by JeffBezanson** [comment]
Excellent arguments, @lindahua . I'd prefer to write `A + λ*I` explicitly for that case.
What do you think about `1/A`?

**Comment by stevengj** [comment]
With regards to `^`, it would be very strange if `A^2` did not equal `A*A`, so I think using the matrix exponentiation operation makes sense there.

**Comment by lindahua** [comment]
The `λI` matrix is a very important one that are widely used in practice. It might be worth introducing special facility to deal with it in an efficient way. @carlobaldassi's identity matrix solution sounds a nice solution to me. With that kind of matrix, one can write `A + λI`, `λI / A`, and so on. It is clean, explicit, and convenient.

For these three operations: `*`, `/`, `\`, and `^`, I think it would be good to have algebraic semantics (for matrices of proper size), while one can use `.*`, `./`, `.\`, and `.^` for element-wise semantics.

`α * A` and `A * α` means the same thing in both sense (in linear algebra, it is called scalar product). 
`A ./ α` and `A * inv(α I)` means the same thing. So there won't be confusion about the expression `A / α`. 

`α / A` is less obvious -- `αI / A` clearly means different things than `α ./ A`. I don't have strong opinion here. Either way is fine.

Personally, I think a preferable way is to make ambiguous things explicit:
- Introduce the `IdentityMatrix` type (or whatever name) to represent `αI`.
- Use `α ./ A` for element-wise operation.
- Use `α I / A`  to express `α * inv(A)` and 
- Deprecate the use of `α / A`.

**Comment by kmsquire** [comment]
+1 for the @carlobaldassi's `IdentityMatrix` idea, with the caveat that I'm sure `I` is being used somewhere as a variable. :-/

**Comment by lindahua** [comment]
A local variable `I` would hide the constant identity matrix `I`. So probably won't break existing codes. (Just like `e` is also being used as local variables/fields in many places).

If we add this constant `I`, it will only be used in new/revised codes. In such cases, I believe most people would avoid using `I` as a local variable whenever he wants to use `I` as an identity matrix within the same local context.

**Comment by milktrader** [comment]
@lindahua thanks for the excellent explanation

**Comment by stevengj** [comment]
I like the `IdentityMatrix` idea too.

One could then define `I[n,m]` as a Kronecker delta function as well.

However, I don't think `IdentityMatrix` is the right name for the type, since it includes a scale factor.  Maybe `ScalingOperator` (since it could be an operator for arbitrary vector space, and in any case is not a `<: AbstractMatrix` since it doesn't have a `size`)

**Comment by carlobaldassi** [comment]
I agree with @lindahua on everything, included the preference for deprecating `α/I` and forcing to be explicit on that.
But I'm actually a bit (not too much really, but still) concerned about using `I` as a constant; it's true it would not probably break any code, but a quick grepping in Julia base reveals it's very commonly used (480 hits), especially to indicate a group of indices. The best choice would probably be one of `\U1D7D9` (['MATHEMATICAL DOUBLE-STRUCK DIGIT ONE'](http://www.fileformat.info/info/unicode/char/1d7d9/index.htm)) or `\U1D540` (['MATHEMATICAL DOUBLE-STRUCK CAPITAL I'](http://www.fileformat.info/info/unicode/char/1d540/index.htm)); maybe the (admittedly ugly) name `II` could be used to mimic that. Or, following `ⅈ` → `im` it could be called `ID`. So in Base code we would use Unicode, but users could still use ASCII without too much hassle.

**Comment by nalimilan** [comment]
@carlobaldassi Unfortunately, the standard notation for the identity matrix really seems to be `I`. Even though `1` is used sometimes, it is also used to designate a vector/matrix of ones, so that's not ideal. And I with double strike does not seem to be used anywhere.

**Comment by StefanKarpinski** [comment]
Using `I` seems fine to me since it won't break any existing code. The same name means different things in different contexts. That's just normal and not really a big deal at all.

**Comment by andreasnoack** [comment]
Sorry for the long break. I can see that you have had a great party here. 

I am not proposing the following behaviour in julia, but just for completeness, here is [Sage](http://www.sagemath.org)

``` sage
A=matrix([[1,1],[1,1]])

A+2
[3 1]
[1 3]
```

I think it will be better to add the `IdentityMatrix`.

Generally, I think that it would be cleaner to let the `.` versions do the less-mathy elementwise operations and reserve `+`,`-`,`*`,`/`,`^` for the math operations, i.e. make `randn(2,2)+1` and error and point to `.+`. I don't see the benefit of sometimes having the non dot version default to the dot version.

Finally, even though we might keep `+` and `-` to default to elemenwise operations, I don't see why the fact that `A^-1=inv(A)` should imply that `1/A=inv(A)`. I think the `Number/Array` methods should go away except maybe for the case mentioned earlier, but that exception is not so important to me.

**Comment by JeffBezanson** [comment]
I'd be fine with removing Number/Array.

**Comment by lsorber** [comment]
+1 for @andreasnoackjensen's arguments

**Comment by toivoh** [comment]
I want to put in my vote for letting `matrix+scalar` give an error, or be allowed only for square matrices where it would give `matrix+scalar*I`.

It was not always obvious that Julia should have both `+` and `.+` operators. After all, `+` would typically just be a more restricted form of the broadcasting `.+`, so why not let `+` broadcast? The argument that finally swayed Stefan was that then `*` (matrix multiplication) distributes over `+` (non-broadcasting addition), while `.*` distributes over `.+`. This should definitely lead to fewer surprises.

Put another way, when you are writing linear algebra code using `+` and `*`, you really prefer an error when you try to add `scalar+vector`, `scalar+matrix`, or `vector+matrix`, rather than silently getting a broadcasted result. A lot of the strength of the `+` operator compared to `.+` comes from the cases that it does _not_ allow.

When it comes to being explicit about what operations do, I think that `matrix+scalar` working elementwise is really sneaky. If you want a broadcasting operation, be explicit about it and use `.+`. I think that a good way forward be to deprecate `matrix+scalar`, and make everyone use `.+` instead.

If the distributive property should hold for `matrix+scalar` as well, it must evaluate to `matrix+scalar*I`, so that e.g.

```
matrix*vector + scalar*vector = (matrix+scalar)*vector = (matrix+scalar*I)*vector
```

But I think it would be ok if you have to put in the `I` yourself. Cases might come up where the above would be convenient, but it if we deprecate `matrix+scalar`, that can always be discussed later on.

For `1/A` where `A` is a matrix, I want to appeal to associativity. If we want that

```
B*1/A = (B*1)/A = B*(1/A)
```

then we need `1/A = inv(A)`, which I think is the right choice. This also goes in line with letting `+ - * /` be linear algebra operators.

A lot of you have talked about consistency with other languages. I beg of you to consider the internal consistency of Julia as well. For this, we need to consider mathematical arguments. When one gains experience with numerical software, it easy to be lead to believe that it is by necessity much less consistent than mathematics, with corner cases that you have to watch out for everywhere you turn. Experiencing Julia has led me to believe that it does not have be so, it's more that too many programming languages were designed without such considerations, often by people without sufficient mathematical background. I still hope that Julia can be different. I would argue that satisfying basic properties like distributivity and associativity is an important consistency aspect, and generic programming thrives on consistency.

**Comment by andreasnoack** [comment]
@toivoh Thanks for the comment. After reading it twice, I am convinced by your arguments, and also annoyed that I didn't see the associativity argument for `inv(A)=1/A`.

**Comment by Jutho** [comment]
It looks like a lot of the discussion is on the difference between interpreting an Array as just a multidimensional collection of data versus interpreting them as a mathematical object (vector,matrix, tensor = multilinear map).

My vote would also be on having number/matrix = number \* inv(matrix), and on deprecating array + number and requesting users to change the code to array .+ number, despite @lindahua 's argument that this construction is often used in the technical computing community.  I agree with @lindahua 's argument for not having matrix + number adding to the diagonal (cited below), but the same argument also requires not to have matrix + number = adding number everywhere. In fact, I cannot think of any algebraic construction that would ever want to use such a strange operation, but that's probably just my ignorance.

Citing @lindahua 
"From a pure algebraic standpoint, the Matrix Ring is formed of all n x n square matrices together with the addition and matrix multiplication defined thereon. Perhaps, I am missing something. I have yet to see an algebraic system (in terms of mathematical concept) that involves addition between matrix and scalar that exhibit the A + λI behaviour."

**Comment by lindahua** [comment]
I don't have a problem deprecating `Array + Number` in favor of `Array .+ Number`. I am actually against quietly changing the behavior/semantics in a radical way.

If I were asked to design a system, I would prefer such a system:
- Using dot-prefixed operators `.+`, `.-`, `.*`, `./`, `.\`, and `.^` for element-wise operations. 
- Using `+`, `-`, `*`, `\`, and `/` for (linear) algebraic operations between arrays (considered as tensors)
- `A ^ n` should be interpreted as algebraic power, _i.e._ `A * A * A ...`.
- Allow `λ * A` and `A / λ` for scalar product (this is also well defined in linear algebra)
- adding scalars to diagonals should be written as `A + λI`.  `A + λ` and `A - λ` not allowed (due to potential confusion).
- `λ * inv(A)` may be written as `λI / A`. `λ / A` is not allowed.

**Comment by johnmyleswhite** [comment]
+1 for that

**Comment by jiahao** [comment]
+1 in favor of @lindahua's proposal.

`1/A` as matrix inverse is cute, but more likely symptomatic of someone who doesn't quite know what they're doing. And people really should avoid computing inverses explicitly anyway.

**Comment by StefanKarpinski** [comment]
Yes, that seems like a very sane approach. +1

**Comment by simleb** [comment]
+1

**Comment by andreasnoack** [comment]
I support @lindahua's proposal except that I think @toivoh's argument for allowing `1/A` is convincing. Now that @jiahao raises the often heard warning against the inverse, I would like to ask you about [Druinsky and Toledo](http://arxiv.org/pdf/1201.6035v1.pdf). I have also made a [plot](http://nbviewer.ipython.org/gist/andreasnoackjensen/9218345).

**Comment by jiahao** [comment]
You've caught me reciting scripture (*￣m￣)

It isn't so much that `inv(A)*b` is less accurate than `A\b` - that obviously depends on the actual algorithms used for `inv` and `\` - , but rather that it is significantly slower in many practical use cases. Three in particular come to mind:
1. For sparse matrices, computing `inv(A)` is often ill-advised, since it usually has a very different sparsity pattern from `A` itself, and is prone to fill-in. So even if `A` is sparse, `inv(A)` usually isn't.
2. For singular matrices, you can graciously overload `\` compute the least-squares solution `A\b`, which `inv(A)*b` won't give you - you'd need `pinv(A)*b` for that, which is significantly more expensive.
3. For use cases with multiple rhs `b`s, it's faster to factorize once and solve with the factorization multiple times, as opposed to inverting the matrix once and computing multiple matvecs. The factorizations usually involve special matrix forms for which the matvecs would be faster.

You'll notice in your notebook that the spread of errors using `A\b` is narrower than for `inv(A)*b`. I think this is indicative of doing more flops in the latter than the former, and that you can have cancellation pushing the error lower, or componding error pushing the error higher. I don't think that is necessarily surprising.

I really can't think of a compelling use case for matrix inverses where the right to do is to express it using `1/A`s.

**Comment by kmsquire** [comment]
> You'll notice in your notebook that the spread of errors using A\b is narrower than for inv(A)*b

Actually, I think it's the other way around.  LU errors are circles, INV errors are x's.

**Comment by jiahao** [comment]
Oops, fail.

I'm not sure why that would be the case. I notice that inv and \ are both using the same LU factorization under the hood, is that right?

**Comment by toivoh** [comment]
+1 to @lindahua's suggestion, explicit is better than implicit. My argument for 1/A = inv (A) should perhaps even more be taken as argument against elementwise division.

I think that we should consider what `I` really is though. I would see it as an (identity) scaling operator, something more general than an identity matrix with unknown size.

**Comment by Jutho** [comment]
+1 for @lindahua's suggestion. I had no strong preference for 1/A being either inv(A) or just an error (but not elementwise division), and would definitely not use the 1/A notation myself.

**Comment by andreasnoack** [comment]
I have updated the pull request JuliaLang/julia#5810 in consequence of the discussion here.

@jiahao I understand that there are many situations where the `\` version is preferred, but I think it interesting that the broadly communicated warning against the numeric accuracy of `inv(A)*b` doesn't really hold. I have updated the [plot](http://nbviewer.ipython.org/gist/andreasnoackjensen/9218345) with relative residual/backward errors. It explains the warning: `inv(A)*b` is not backward stable in `b`, but it must be the error of the solution that matters. One of the [cited papers](http://link.springer.com/article/10.1007/s10915-006-9112-x) points out that the only place in Higham's 14 pages on matrix inversion where the forward error of `inv(A)*b` is mentioned is in an exercise where you asked to derive and interpret the forward error. That paper also gives examples where use of `inv(A)*b` is faster because of repeated use `inv(A)` for changing `b`s. I can't tell if their examples are reasonable or their timings fair.

**Comment by stevengj** [comment]
`inv(A)*b`, given `inv(A)`, requires asymptotically the same number (2m² for an m×m matrix) of flops as the two backsubstitutions in `U \ L \ b`, given the LU factors of A.    The practical constant factor in the real execution time of the former is slightly better because matrix-vector multiplication has greater regularity than backsubstitution, I think, and hence is easier to optimize.   However, the difference only matters if you have a huge number of right hand sides ( >> m), as otherwise the time to compute the inverse dominates.  But I agree that it is mostly a question of performance, not accuracy.

(In practice, I suspect that inverting matrices for performance reasons is mostly seen for tiny matrices, e.g. 3×3. On the other hand, for pedagogical and exploration purposes, I compute inverse matrices of moderate-sized matrices all the time.)

**Comment by JeffBezanson** [comment]
closed by JuliaLang/julia#5810
